core/src/org/purefn/kurosawa/config/env.clj

```diff
@@ -0,0 +1,39 @@
+(ns org.purefn.kurosawa.config.env
+  "Load configuration from environment variables."
+  (:require [clojure.string :as str]
+            [org.purefn.kurosawa.config.parse :as parse]))
+
+(defn- munge-name
+  [s]
+  (str/join
+   "-"
+   (-> (str/lower-case s)
+       (str/split #"_")
+       (rest))))
+
+(defn- env-variables*
+  []
+  (->> (System/getenv)
+       (map (juxt (comp str/lower-case key)
+                  (comp munge-name key)
+                  val))
+       (group-by (comp first #(str/split % #"_") first))
+       (into {})))
+
+(def env-variables (memoize env-variables*))
+
+(defn fetch-config
```

Comment:
So in terms of design, you're going with a pattern of grabbing pieces of the config by name. Did you consider an approach of "give me the whole config as a map"?

Reply:
We're constrained to this contract (for now) until we can refactor nearly every `defrecord` constructor in our codebase's usage of the [k8s namespace](https://github.com/PureFnOrg/kurosawa/blob/config/core/src/org/purefn/kurosawa/k8s.clj).

That doesn't mean we couldn't load the whole config at once though, and statefully store it in an `atom`, and each component fetching a piece.

Reply:
Question is, what is the ideal state of affairs? Personally I think the right way to do config is have it all in the service, the service does whatever it needs to do to get a big map of all the config, passes that config map into an init function that creates the system map with all the configured components, and then starts that.

I don't think that having components be responsible for their own config is the ideal state of affairs. So the larger question is, which do you think is the ideal state of affairs?

If you agree with me, then I think steps can be taken without requiring an immediate rewrite of these components to get closer to that state of affairs. Basically what you're suggesting seems pretty good, but with the idea of that global atom being a transitional, deprecated way of going about it to support the current access patterns of components. Maybe something in the `config` namespace like an `init-config` function that sets the atom, but just sets it to whatever is passed in, doesn't try to go fetching from anywhere. That leaves the responsibility of building that config map up to the services, and there should be various utility functions (like the ones in this PR) to use to do that.

Reply:
In general I agree, with the caveat that a component should be able to coerce an input map of config into the structure it expects, then validate it against a spec.  Otherwise a whole lot of copy-paste goes on between deployments, jobs, streams apps, dev namespaces, etc.

Of course you can always wire up configuration manually, but all of the import sources outlined here will emit a config map of the same structure, 2-levels deep, with string, kebab-cased, keys.

I'm also still trying to get to a place where something like
```(create-system [:big-domain-component])``` is possible, injected with all dependencies and config from the current environment.  So wherever config lands should support that too, ultimately.

Reply:
I don't think it makes sense to encode knowledge about how config is going to be represented in the underlying config storage mechanism into the components. They shouldn't need to know how to turn ssm/environment vars/whatever into the specific structure they want. The separation of concerns should have that knowledge in the thing that's handling config - somewhere in the service.